File: LeetCode/src/dp/BestTimeToBuyAndSellStockIV.java
package dp;

public class BestTimeToBuyAndSellStockIV {
    public int maxProfit(int k, int[] prices) {
        int n = prices.length;
        if (n == 0 || k == 0) return 0;
        if (k >= n / 2) return maxProfitUnlimitedTransactions(prices);

        int[][] dp = new int[k + 1][n];
        for (int i = 1; i <= k; i++) {
            int maxDiff = -prices[0];
            for (int j = 1; j < n; j++) {
                dp[i][j] = Math.max(dp[i][j - 1], maxDiff + prices[j]);
                maxDiff = Math.max(maxDiff, dp[i - 1][j] - prices[j]);
            }
        }
        return dp[k][n - 1];
    }

    private int maxProfitUnlimitedTransactions(int[] prices) {
        int maxProfit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1]) {
                maxProfit += prices[i] - prices[i - 1];
            }
        }
        return maxProfit;
    }

    public static void main(String[] args) {
        BestTimeToBuyAndSellStockIV solution = new BestTimeToBuyAndSellStockIV();

        // Test case
        int k = 2;
        int[] prices = {2,4,1};
        System.out.println(solution.maxProfit(k, prices)); // Output: 2
    }
}

File: LeetCode_Python/dp/stock/BestTimeToBuyAndSellStockIV.py
class BestTimeToBuyAndSellStockIV:
    def maxProfit(self, k: int, prices: list[int]) -> int:
        n = len(prices)

        if k >= n // 2:
            profit = 0
            for i in range(1, n):
                if prices[i] > prices[i - 1]:
                    profit += prices[i] - prices[i - 1]
            return profit

        total = 2 * k + 1
        dp = [[0] * total for _ in range(n)]

        for j in range(1, total):
            if j % 2 == 1:
                dp[0][j] = -prices[0]

        for i in range(1, n):
            for j in range(1, total):
                if j % 2 == 0:
                    dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - 1] + prices[i])
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - 1] - prices[i])

        return max(dp[-1][j] for j in range(0, total, 2))


# Time Complexity: O(n * k), where n is the number of prices and k is the number of transactions
# Space Complexity: O(n * k), for the dp array

if __name__ == "__main__":
    solution = BestTimeToBuyAndSellStockIV()

    print("Expected Output: 2")
    print("Actual Output:", solution.maxProfit(2, [2, 4, 1]))

    print("Expected Output: 7")
    print("Actual Output:", solution.maxProfit(2, [3, 2, 6, 5, 0, 3]))

    print("Expected Output: 0")
    print("Actual Output:", solution.maxProfit(0, [1, 2, 3]))
